oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java

```diff
@@ -0,0 +1,173 @@
+import org.jclouds.collect.Memoized;
+import org.jclouds.compute.domain.Hardware;
+import org.jclouds.compute.domain.HardwareBuilder;
+import org.jclouds.compute.domain.NodeMetadata;
+import org.jclouds.compute.domain.NodeMetadataBuilder;
+import org.jclouds.compute.domain.OperatingSystem;
+import org.jclouds.compute.domain.OsFamily;
+import org.jclouds.compute.domain.Processor;
+import org.jclouds.compute.domain.Volume;
+import org.jclouds.compute.functions.GroupNamingConvention;
+import org.jclouds.domain.Location;
+import static org.jclouds.location.predicates.LocationPredicates.idEquals;
+import org.jclouds.util.InetAddresses2;
+
+public class ServerToNodeMetadata implements Function<Server, NodeMetadata> {
+
+   private final Supplier<Set<? extends Location>> locations;
+   private final Function<Hdd, Volume> fnVolume;
+   private final Function<List<ServerIp>, List<String>> fnCollectIps;
+   private final OneAndOneApi api;
+
+   private final GroupNamingConvention groupNamingConvention;
+
+   @Inject
+   public ServerToNodeMetadata(Function<Hdd, Volume> fnVolume,
+           @Memoized Supplier<Set<? extends Location>> locations,
+           OneAndOneApi api,
+           GroupNamingConvention.Factory groupNamingConvention) {
+      this.locations = locations;
+      this.api = api;
+      this.fnVolume = fnVolume;
+      this.fnCollectIps = new Function<List<ServerIp>, List<String>>() {
+         @Override
+         public List<String> apply(List<ServerIp> in) {
+            List<String> ips = Lists.newArrayListWithExpectedSize(in.size());
+            for (ServerIp ip : in) {
+               ips.add(ip.ip());
+            }
+            return ips;
+         }
+      };
+      this.groupNamingConvention = groupNamingConvention.createWithoutPrefix();
+   }
+
+   @Override
+   public NodeMetadata apply(final Server server) {
+      checkNotNull(server, "Null server");
+
+      DataCenter dataCenter = api.dataCenterApi().get(server.datacenter().id());
+      Location location = find(locations.get(), idEquals(dataCenter.location()));
+
+      double size = 0d;
+      List<Volume> volumes = Lists.newArrayList();
+      List<Hdd> storages = server.hardware().hdds();
+      if (storages != null) {
+         for (Hdd storage : storages) {
+            size += storage.size();
+            volumes.add(fnVolume.apply(storage));
+         }
+      }
+
+      // Build hardware
+      String id = String.format("cpu=%f,ram=%d,disk=%f", server.hardware().coresPerProcessor(), (int) server.hardware().ram(), size);
+      Hardware hardware = new HardwareBuilder()
+              .ids(id)
+              .name(id)
+              .ram((int) server.hardware().ram())
+              .processor(new Processor(server.hardware().coresPerProcessor(), 1d))
+              .hypervisor("kvm")
+              .volumes(volumes)
+              .location(location)
+              .build();
+
+      // Collect ips
+      List<String> addresses = fnCollectIps.apply(server.ips());

```

Comment:
No need for that dummy function in the constructor. Remove it and change this to something like:

```java
List<String> addresses = Lists.transform(new Function<ServerIp, String>() {
   @Override public String apply(ServerIp in) {
      return in.ip();
   }
});
```